### PR TITLE
feat(cdk/dialog, cdk/overlay): support open and close animations for overlay and backdrop

### LINKED ISSUE #28878: feat(cdk/dialog): Support for Leaving Animation - wait for all child animations to complete before disposing the overlay
### Feature Description

Currently, when closing a dialog, the corresponding overlay is immediately disposed (L106):

https://github.com/angular/components/blob/e3eef47f80a7c1dda4b538128fed7f530f7cdff1/src/cdk/dialog/dialog-ref.ts#L99-L113

And when disposing an overlay, the host `<div>` element of the overlay is immediately removed from the DOM (L269):

https://github.com/angular/components/blob/e3eef47f80a7c1dda4b538128fed7f530f7cdff1/src/cdk/overlay/overlay-ref.ts#L252-L280

The current behavior completely eliminated the possibility to implement a `:leave` animation for dialogs, since the ancestor element of the dialog, the overlay, is removed via the native DOM API, and the Angular animation engine cannot hook into this process to perform the `:leave` animation.

As a solution, I propose to allow the DialogContainer to decide when the overlay should be disposed. By doing so, we can create another DialogContainer aside from CdkDialogContainer, e.g. LeaveAnimationAwareDialogContainer, which performs a leave animation that does nothing but `animateChild()`. The LeaveAnimationAwareDialogContainer notifies the DialogRef to dispose the overlay after the leaving animation is completed.

### Use Case

_No response_

### PR DESCRIPTION
**Pull request by NickUfer**
This PR removes the CSS based opening animations for the backdrop of cdk/overlay and replaces it with the options to define opening and closing angular animations for the cdk/overlay overlay and backdrop.

Closing animations are currently not possible as the disposal does not wait for animations to finish, this PR fixes this and adds more flexibility and ease of use.

It also adds default animations for all cases for the cdk/dialog, but disables them for the material ui dialog to prevent any interference with the custom material ui dialog animations.

Fixes #28878 

Demo:

https://github.com/angular/components/assets/11254522/6561b60d-2d12-4fac-9115-af164ccd7194

### CODE REVIEW

**Comment by NickUfer** [comment]
Will fix and add new tests once I know if this can even be merged.

**Review by crisbeto — COMMENTED** [review]
We're generally trying to move away from `@angular/animations` in favor of defining the animations directly with CSS so I don't think that we'd want to introduce any new APIs based around it.

**Comment by NickUfer** [comment]
Hmm, that looks to me like a bad idea to be honest. But if thats the way, ok.

**Comment by crisbeto** [comment]
To elaborate on why we want to move away from it long-term:
1. Our animations are fairly basic so we don't have much out of the module in the first place.
2. We don't want to force our users to install the module.
3. The animations DSL can be a bit hard to follow which makes the animations harder to change.
4. Over the years there have been issues with memory leaks coming specifically from the animations module.

**Comment by NickUfer** [comment]
Thanks for the insight!

> 1. Our animations are fairly basic so we don't have much out of the module in the first place.

Makes sense to me for the material ui as it is opinionated and it is expected that people do not want to change it, but that does not really work for CDK in my opinion. The CDK is described in the README in the following way:

> @angular/cdk: Library that helps you author custom UI components with common interaction patterns

---

> 3. The animations DSL can be a bit hard to follow which makes the animations harder to change.

Subjective, but I see where you are coming from. For me personally it is much easier and flexible than defining it in CSS.

---

2 and 4, yeah, valid... 4 could be fixed though but probably unrealistic.
